### PR TITLE
Sync GitHub workflows from main branch

### PR DESCRIPTION
**Pull request by prudhvigodithi**
### Description
While working on this failed backport https://github.com/opensearch-project/OpenSearch/pull/13786#issuecomment-2125934247, I have noticed majority of the existing workflows are either not backported, synced or updated, I have tried to find what ever I can and updated through this PR.

### Related Issues
Part of: https://github.com/opensearch-project/OpenSearch/issues/3713

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
I see the same reason why the gradle checks are failing.
https://github.com/opensearch-project/OpenSearch/pull/13787#issuecomment-2125908567

File: .github/workflows/gradle-check.yml [hunk 1]
```diff
@@ -62,28 +96,58 @@ jobs:
...
         with:
           issue-number: ${{ env.pr_number }}
           body: |
-              ### Gradle Check (Jenkins) Run Completed with:
-              * **RESULT:** ${{ env.result }} :x:
-              * **URL:** ${{ env.workflow_url }}
-              * **CommitID:** ${{ env.pr_from_sha }}
+            :x: Gradle check result for ${{ env.pr_from_sha }}: [${{ env.result }}](${{ env.workflow_url }})
+
+            Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
```
Comment (reta) [inline .github/workflows/gradle-check.yml:144]: I think `2.x` would be more appropriate here:

```suggestion
            Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/2.x/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
```

Comment (prudhvigodithi) [reply .github/workflows/gradle-check.yml]: Thanks @reta, just updated in my latest commit.

**Comment by reta** [comment]
@prudhvigodithi do you need to rabase? (incorporating https://github.com/opensearch-project/OpenSearch/pull/13787#issuecomment-2125908567)

**Comment by prudhvigodithi** [comment]
> @prudhvigodithi do you need to rabase? (incorporating [#13787 (comment)](https://github.com/opensearch-project/OpenSearch/pull/13787#issuecomment-2125908567))

This has to be force merged @reta and existing PR's (to execute gradle check) should be rebased.
Thanks

**Comment by reta** [comment]
> This has to be force merged @reta and existing PR's (to execute gradle check) should be rebased.



> This has to be force merged @reta and existing PR's (to execute gradle check) should be rebased.

The lastest build still fails with:

```
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: f9e7ab7d-86da-45c5-94f4-d8c8729de4f7
groovy.lang.MissingPropertyException: No such property: post_merge_action for class: groovy.lang.Binding
	at groovy.lang.Binding.getVariable(Binding.java:63)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onGetProperty(SandboxInterceptor.java:285)
	at org.kohsuke.groovy.sandbox.impl.Checker$7.call(Checker.java:375)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:379)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
```

**Comment by prudhvigodithi** [comment]
> The lastest build still fails with:

Ya thats because I'm trying to test (re-run the gradle check) with old open PR's to see if there is any way to avoid rebase. For any PR's that created after this was merged (example on `main` and `2.x` line https://github.com/opensearch-project/OpenSearch/pull/13793) will have no issues with the gradle check and jenkins build.

File: .github/workflows/assemble.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+name: Gradle Assemble
+on: [pull_request]
+
+jobs:
+  assemble:
+    if: github.repository == 'opensearch-project/OpenSearch'
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        java: [ 11 ]
```
Comment (reta) [inline .github/workflows/assemble.yml:10]: I believe we maintain JDK-8 compatibility in 1.x:

```suggestion
        java: [ 8, 11 ]
```

File: .github/workflows/precommit.yml [hunk 3]
```diff
@@ -3,14 +3,20 @@ on: [pull_request]
 
 jobs:
   precommit:
-    runs-on: ubuntu-latest
+    if: github.repository == 'opensearch-project/OpenSearch'
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        java: [ 11 ]
```
Comment (reta) [inline .github/workflows/precommit.yml:10]:
```suggestion
        java: [ 8, 11 ]
```

**Comment by reta** [comment]
@prudhvigodithi sorry about that:

```
* Where:
Build file '/home/runner/work/OpenSearch/OpenSearch/buildSrc/build.gradle' line: 78
```

it seems like 8 is no go :( my apologies

**Comment by prudhvigodithi** [comment]
Updated @reta, also for assemble task to work on windows, do we need to backport this PR https://github.com/opensearch-project/OpenSearch/pull/1412 ? As I see the run passed https://github.com/opensearch-project/OpenSearch/actions/runs/9181040634/job/25246840531?pr=13754 skipped the following tasks
```
:distribution:docker:buildArm64DockerImage
:distribution:docker:buildDockerImage
:distribution:docker:buildPpc64leDockerImage
```
@dblock @bbarani

**Comment by prudhvigodithi** [comment]
Also FYI @bbarani @dblock @reta  for now I have created 2 new branch protection rules for `1.3` and `1.x` (once we have the finalized list we have a regex as `1.*`), this should take care of only checks limited to only `1.x` series. Coming to the `Minimum approval count` and `copy-issue-labels`, the conditions used are `pull_request_review` and `pull_request_target`, so these workflows expects a workflow file from source repo and not fork and they should start working once we merge this PR.

**Comment by prudhvigodithi** [comment]
> Updated to only 11 @reta, also for assemble task to work on windows, do we need to backport this PR #1412 ? As I see the run passed https://github.com/opensearch-project/OpenSearch/actions/runs/9181040634/job/25246840531?pr=13754 skipped the following tasks
> 
> ```
> :distribution:docker:buildArm64DockerImage
> :distribution:docker:buildDockerImage
> :distribution:docker:buildPpc64leDockerImage
> ```
> 
> @dblock @bbarani

I was able to handle this within the `assemble` workflow, the `assemble` workflows is green now.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
Chatted with @prudhvigodithi , the Gradle check fails with because the original branch (1.x) does not have the ENV variable populated. 

```
Error when executing always post condition:
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 362f15cc-7542-4a54-a0b6-4af2707ce3d5
groovy.lang.MissingPropertyException: No such property: post_merge_action for class: groovy.lang.Binding
	at groovy.lang.Binding.getVariable(Binding.java:63)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onGetProperty(SandboxInterceptor.java:285)
	at org.kohsuke.groovy.sandbox.impl.Checker$7.call(Checker.java:375)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:379)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:355)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.getProperty(SandboxInvoker.java:29)
	at com.cloudbees.groovy.cps.impl.PropertyAccessBlock.rawGet(PropertyAccessBlock.java:20)
	at WorkflowScript.run(WorkflowScript:172)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.get(PropertyishBlock.java:73)
	at com.cloudbees.groovy.cps.LValueBlock$GetAdapter.receive(LValueBlock.java:30)
```

**Comment by prudhvigodithi** [comment]
Thanks @reta will wait for a backport PR for 1.3, else I will create it manually.

**Comment by prudhvigodithi** [comment]
Added the backport workflow but since this still failed to backport raised a manual backport PR https://github.com/opensearch-project/OpenSearch/pull/13804.

**Pull request merged**